homeassistant/components/upnp/config_flow.py

```diff
@@ -1,10 +1,172 @@
 """Config flow for UPNP."""
+from typing import Mapping, Optional
+import voluptuous as vol
+
 from homeassistant import config_entries
-from homeassistant.helpers import config_entry_flow
+from homeassistant.components import ssdp
 
-from .const import DOMAIN
+from .const import (  # pylint: disable=unused-import
+    DOMAIN,
+    LOGGER as _LOGGER,
+)
 from .device import Device
+from urllib.parse import urlparse
 
-config_entry_flow.register_discovery_flow(
-    DOMAIN, "UPnP/IGD", Device.async_discover, config_entries.CONN_CLASS_LOCAL_POLL
-)
+
+LOCATION = "location"
+NAME = "name"
+ST = "st"
+UDN = "udn"
+
+
+class UpnpFlowHandler(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a UPnP/IGD config flow."""
+
+    VERSION = 1
+    CONNECTION_CLASS = config_entries.CONN_CLASS_LOCAL_POLL
+
+    # Paths:
+    # - ssdp(discovery_info) --> ssdp_confirm(None) --> ssdp_confirm({}) --> create_entry()
+    # - user(None): scan --> user({...}) --> create_entry()
+    # - import(None) --> create_entry()
+
+    def __init__(self):
+        """Initialize the UPnP/IGD config flow."""
+        self._data: Mapping = None
+        self._discoveries: Mapping = None
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow start."""
+        _LOGGER.debug("async_step_user: user_input: %s", user_input)
+
+        if user_input is not None:
+            # Ensure wanted device was discovered.
+            matching_discoveries = [
+                discovery
+                for discovery in self._discoveries
+                if discovery["unique_id"] == user_input["unique_id"]
+            ]
```

Comment:
This is already validated by the `data_schema` below.

Reply:
This is looking up the discovery-entry with the given unique_id. From this entry, we use the `udn`/`st` (separate values) to create the config entry.